lib/AST/ProtocolConformance.cpp

```diff
@@ -1367,3 +1368,46 @@ ProtocolConformanceRef::getCanonicalConformanceRef() const {
     return *this;
   return ProtocolConformanceRef(getConcrete()->getCanonicalConformance());
 }
+
+// See swift/Basic/Statistic.h for declaration: this enables tracing
+// ProtocolConformances, is defined here to avoid too much layering violation /
+// circular linkage dependency.
+
+struct ProtocolConformanceTraceFormatter
+    : public UnifiedStatsReporter::TraceFormatter {
+  void traceName(const void *Entity, raw_ostream &OS) const {
+    if (!Entity)
+      return;
+    const ProtocolConformance *C =
+        static_cast<const ProtocolConformance *>(Entity);
+    OS << "<conformance ";
+    if (auto const *DC = C->getDeclContext()) {
+      if (auto const *N = DC->getAsNominalTypeOrNominalTypeExtensionContext()) {
+        N->getFullName().print(OS, false);
+      }
+    }
+    OS << " : ";
+    if (auto *P = C->getProtocol())
+      P->getFullName().print(OS, false);
+    OS << ">";
+  }
+  void traceLoc(const void *Entity, SourceManager *SM,
+                clang::SourceManager *CSM, raw_ostream &OS) const {
+    if (!Entity)
+      return;
+    const ProtocolConformance *C =
+        static_cast<const ProtocolConformance *>(Entity);
+    if (auto const *DC = C->getDeclContext()) {
+      if (auto const *D = DC->getAsDeclOrDeclExtensionContext())
+        D->getSourceRange().print(OS, *SM, false);
```

Comment:
I don't think you want the range here, just the location. (Arguably if you have a NormalProtocolConformance you could get an even better location, which is where the conformance is written. But not all conformances are NormalProtocolConformances.)

Reply:
Hm, I know the method is named traceLoc but so far I've been actually tracing all of the ranges. Maybe I should rename it? Or maybe ranges are overkill?

(Will use the normalProtocolConformance decl site when available, in any case)

Reply:
That might be reasonable for some of the other things, but for conformances you'll end up using the entire range of the extension or decl that declares conformance, or maybe just the class that inherits from a conforming decl. That doesn't seem useful.

Reply:
Ok (I think I initially went with ranges because of exprs, but *shrug* it's pretty much just a hint for the reader anyway).